README.md
php-sdk2
========

php-sdk2 is a PHP library that makes the Amazon AWS SimpleDB REST API available
as PHP methods. All 
[API operations](http://docs.amazonwebservices.com/AmazonSimpleDB/latest/DeveloperGuide/SDB_API_Operations.html)
are available:

*    BatchDeleteAttributes
*    BatchPutAttributes
*    CreateDomain
*    DeleteAttributes
*    DeleteDomain
*    DomainMetadata
*    GetAttributes
*    ListDomains
*    PutAttributes
*    Select

Some additional functionality is available:

*    Default handler for service temporarily unavailable, which retries four times with different waiting intervals
     before giving up
*    A total machine hours counter, that adds all box usage returns over the life time of the SimpleDB object
*    A queue PutAttributes and a queue deleteAttributes feature, because one batch request reduces round trips and 
     latencies compared to many single requests. 
     requests. The methods take care themselves to send data when the maximum item number for a batch job is reached.
*    The select method has an additional argument to return the whole result at once. No more NextTokens ;-) (if a large 
     result set is returned, this may need a lot of memory, be careful). 
*    The class has now an adaptable error handling. It can raise PHP warnings, errors, throw Exceptions or do nothing.



This project is a fork of the sourceforge [php-sdk project](https://sourceforge.net/projects/php-sdb/) 
from [Dan Myers(heronblademastr)](https://sourceforge.net/users/heronblademastr).
I forked it because it doesn't seem to be maintained anymore, and some functionality is missing 
(batch_delete, ...).





**Attention**: This class is not compatible with the original class. Find below the incompatibilities:

*    The first argument of the select method has been dropped, because it needed to be empty to work.
*    The select methods returns items as an array of name => attributes, and not as an array of 
     ('name' => name, 'attributes' => attributes)
*    The query and queryWithAttributes methods have been removed, because they are deprecated.
*    If the condition of a conditional putAttributes is not met, no error is triggered 
     (SimpleDB error codes 'ConditionalCheckFailed' and 'AttributeDoesNotExist')
*    batchPutAttributes: The items are exprected in the form $items = array('name1' => array(...), 'name2' => array(...)),
     and not like $items = array(array('name' => 'name1', 'attributes' => array(...)), ...) in the old version.

Install
-------

Get a copy. You have three choices:

1. Click on the nice button where it says "ZIP" on [https://github.com/g-g/php-sdb2](https://github.com/g-g/php-sdb2).
2. Use git to clone the repository.
   `git clone https://github.com/g-g/php-sdb2 /your/directory/path`
3. Check it out with subversion.
   `svn checkout http://svn.github.com/g-g/php-sdb2.git`

If you want to use the library in you own project, you will only need the files in the src directory.

Usage
-----

A basic example

```php
<?php
// Import the SimpleDB classes into your name space
use g_g\php_sdb2\SimpleDB;
use g_g\php_sdb2\SimpleDBException;
require_once('g_g\php_sdb2\sdb2.php');

$simpleDB = new SimpleDB('AWS_KEY', 'AWS_SECRET_KEY');
$simpleDB->setErrorHandling(SimpleDB::ERROR_HANDLING_THROW_EXCEPTION);

try {
    // Let's create a domain
    $domain = 'MYDOMAIN';
    $simpleDB->createDomain($domain);

    // Now let's add an item
    $simpleDB->putAttribute($domain, 'ITEM1', array('a' => 1, 'b' => 3, 'c' => 17));

    // Some more items
    $simpleDB->putAttribute($domain, 'ITEM2', array('a' => 2, 'b' => 4, 'c' => 16));
    $simpleDB->putAttribute($domain, 'ITEM3', array('a' => 3, 'b' => 5, 'c' => 15));

    // and let's query for some items
    // Don't forget to quote the values 
    // (http://docs.amazonwebservices.com/AmazonSimpleDB/latest/DeveloperGuide/QuotingRulesSelect.html)
    $results = $simpleDB->select("select * from `$domain` where a > '1'");
    print_r($results);

    // Clean up!
    // This isn't strictly necessary, because when we delete the domain, all data is lost anyway
    // Don't forget to quote the values 
    // (http://docs.amazonwebservices.com/AmazonSimpleDB/latest/DeveloperGuide/QuotingRulesSelect.html)
    // Basically this works like DELETE * FROM $domain WHERE a > 0
    $simpleDB->deleteWhere($domain, "a > '0'");
    $simpleDB->deleteDomain($domain);
} catch (SimpleDBException $e) {
    echo "Something went wrong: " . $e->getMessage();
}

// Was it expensive?
echo 'This example used up ' . $simpleDB->getTotalBoxUsage() . ' machine hours.';
?>
```

You can find more examples in the [examples directory](https://github.com/g-g/php-sdb2/blob/master/examples).

There are some sample programs for the original library, which work also for this library, except the select and
batchPutAttributes method. Check out the [Webmaster In Residence](http://webmasterinresidence.ca/webmasterinresidence/?p=464)

You may also check the unit tests in SimpleDBTest. There you can find many examples how to use a method, and the 
getResultData() method provides the parameters that are send to AWS SimpleDB and what it returns in each case.

About
=====

Requirements
------------

- PHP >= 5.3
- [optional] PHPUnit 3.5+ to execute the test suite

Submitting bugs and feature requests
------------------------------------

Bugs and feature request are tracked on [GitHub](https://github.com/g-g/php-sdb2/issues).

PRs are welcome ;-)

Authors, copyright and license
------------------------------

php-sdb2 is licensed under a modified BSD license.
See the [LICENSE file](https://github.com/g-g/php-sdb2/blob/master/LICENSE) for details on the license and copyrights.



changelog.txt
0.4.0 (initial checkin 2009-2-17)
 - Supports CreateDomain, DeleteDomain, ListDomains
 - Supports PutAttributes, DeleteAttributes, GetAttributes
 - Supports Query
 - Minimal error reporting (only reports whether there was an error)

0.4.1 (committed 2009-02-18)
 - Support QueryWithAttributes

0.4.2 (committed 2009-02-18, folded into 0.4.3)
 - Support DomainMetadata

0.4.3 (committed 2009-02-18)
 - Allow SSL to be disabled at developer's discretion (removed pre-0.4)
 - Allow VERIFYHOST and VERIFYPEER to be disabled

0.4.4 (committed 2009-02-18)
 - Full error reporting

0.4.5 (committed 2009-02-19)
- Remove explicit attribute count range check in putAttributes.
  The endpoint does this, and the limit may change in the future.
- Allow putAttributes to support an array of values for a single key,
  rather than requiring the dev to supply the same key multiple times.

0.4.6 (committed 2009-03-23)
- Fixed QueryWithAttributes to correctly specify multiple attributes
  in the query string and signature.

0.5.0 (committed 2009-04-03)
- Added Select and BatchPutAttributes
- Additions courtesy of Matthew Lanham

0.7.0 (committed 2010-03-01)
- Incorporated changes made by Rich Helms
  - Fixed error in deleteAttributes
  - Update to 2009-04-15 interface
  - Add BoxUsage, RequestId, NextToken, and ErrorCode tracking for the last request
  - Support for ConsistentRead
  - Support for conditional put and conditional delete

0.7.1 (committed 2010-03-25)
- Removed an extraneous, broken debug message (afx114)

0.7.2
- 'expected' parameter to putAttributes and deleteAttributes should be optional (afx114)
- Made __triggerError more informative when curl errors out (afx114)
- Use a modified rawurlencode that's compliant with RFC 3986 as required by AWS. (afx114)

0.7.3
- Corrected checking for 'value's in $expected in putAttribute and deleteAttribute. (afx114)
- Added optional 'host' parameter to SimpleDB constructor.

0.8.0
- The static usage of this class was broken, and has been removed.
- Improved the way the internal SimpleDBRequest object is built.
- Added a callback to support automatic retries after 503 Service Temporarily Unavailable errors.
- Added error checking and a constant for the max batch size in batchPutAttributes.

0.8.1
- Fixed the new defaultReplace flag (added in 0.8.0 but not documented).

0.8.2
- Fixed the way the retry callback was being called.

1.0.beta1
- Major rework
- Unit tests
- Incompatibilities: select and batchPutAttributes have changed.
- New queueDelete and queuePutAttribute methods have been added, see https://github.com/g-g/php-sdb2/examples/queuePutAttributes.php
- A getInstance method provides a singe class for all connections that share aws key and host
- A total box usage feature that sums up the box usage time over the lifetime of the connection object
- Added a batchDeleteAttributes method
- Fixed the service unavailable callback function and provided a default callback
- Internal code refactoring
- The code is now name spaced g_g/php_sdb2
- The error handling has changed. You can choose between exception, user warning, user error or just ignore
- Many others

1.0.rc1
- fixed PHP notice: Undefined property: stdClass::$body


examples/basic_usage.php
<?php
/**
 * This file shows basic php-sdb2 examples 
 * 
 * @package php-sdb2
 *
 * @author Georg Gell
 *
 * @license: For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code, or @link https://github.com/g-g/php-sdb2/blob/master/LICENSE.
 *
 * @link https://github.com/g-g/php-sdb2
 */

// Import the SimpleDB classes into your name space
use g_g\php_sdb2\SimpleDB;
use g_g\php_sdb2\SimpleDBException;
require_once('g_g\php_sdb2\sdb2.php');

$awsKey = '';
$awsSecretKey = '';

if (empty($awsKey) || empty($awsSecretKey)) {
    die('Please set you AWS credentials before running this script');
}

$simpleDB = new SimpleDB($awsKey, $awsSecretKey);
$simpleDB->setErrorHandling(SimpleDB::ERROR_HANDLING_THROW_EXCEPTION);

try {

// Let's create a domain
    $domain = 'MYDOMAIN';
    $simpleDB->createDomain($domain);

// Now let's add an item
    $simpleDB->putAttribute($domain, 'ITEM1', array('a' => 1, 'b' => 3, 'c' => 17));

// Some more items
    $simpleDB->putAttribute($domain, 'ITEM2', array('a' => 2, 'b' => 4, 'c' => 16));
    $simpleDB->putAttribute($domain, 'ITEM3', array('a' => 3, 'b' => 5, 'c' => 15));

// Find some items
// Don't forget to quote the values 
// (http://docs.amazonwebservices.com/AmazonSimpleDB/latest/DeveloperGuide/QuotingRulesSelect.html)
    $results = $simpleDB->select("select * from $domain where a > '1'");
    print_r($results);

// Clean up!
// This isn't strictly necessary, because when we delete the domain, all data is lost anyway
// Don't forget to quote the values 
// (http://docs.amazonwebservices.com/AmazonSimpleDB/latest/DeveloperGuide/QuotingRulesSelect.html)
// Basically this works like DELETE * FROM $domain WHERE a > 0
    $simpleDB->deleteWhere($domain, "a > '0'");
    $simpleDB->deleteDomain($domain);
} catch (SimpleDBException $e) {
    echo "Something went wrong: " . $e->getMessage();
}

// Was it expensive?
echo 'This example used up ' . $simpleDB->getTotalBoxUsage() . ' machine hours.';
?>


?>
